Add `goto-bmc`, a new tool to run only the verifier

NlightNFotis opened the pull request:
This adds a new tool, `goto-bmc`, which you can think of as `CBMC`-lite: its purpose is to run only the backend, on a binary that is produced to be [`symex-ready`](https://diffblue.github.io/cbmc/adr/symex-ready-goto.html), i.e. have been lowered to a form that is accepted by symex without the need for any more transformations.

The aim of the addition of this tool is to contribute towards the efforts to make `CBMC` and all of the assorted tools more modularised, by providing access to the backend of CBMC without needing to bundle frontends, transformation passes, etc, so that usage or integration of the tool into other software or (CI/dev) pipelines becomes easier.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Split API verify_model into two separate steps.

To be able to run the BMC part of CBMC (symex and the
decision-procedure) the API `verify_model` function has been split in
two separate steps: the lowering step that applies all the
transformations to the core-goto program and the verification step that
only runs symex and the decision-procedure.

Commit:
Add goto-bmc executable.

Add a new `goto-bmc` executable to CBMC that reads a core-goto program
from a given binary file and runs symex and the decision procedure.
This has been done by using the API functionalities instead of addin a
new argument to CBMC.

NlightNFotis committed:
Add translation of final_result to CPROVER_EXIT_CODES

NlightNFotis committed:
Add regression tests for basic functionality of goto-bmc

TGWDB commented on src/libcprover-cpp/api.h:
```diff
@@ -98,7 +98,8 @@ struct api_sessiont
   /// Implement necessary transformations to reduce model to core-GOTO, before
   /// being fed to symex.
-  /// @return The function return `true` if it failed because CBMC produced an error.
+  /// @return The function return `true` if it failed because CBMC produced
```
return -> returns

TGWDB commented on src/libcprover-cpp/verification_result.cpp:
```diff
@@ -149,3 +150,21 @@ verification_resultt::get_property_status(const std::string &property_id) const
   UNREACHABLE;
 }
+
+// FOTIS' note: Modelled after `result_to_exit_code` in
+// `src/goto-checker/properties.cpp`.
```
Should this be refactored to use the same code then?

NlightNFotis replied:
We can't do that because the functions use different types as inputs.

We have two types that are equivalent, but one of them is internal to CBMC (`goto-checker`, really), and thus, we can't expose it to the C++ API (because of the internal headers issue), and the other is the C++ API equivalent, which is present in the headers for the API.

We have in the API (under `verification_result.cpp`) a function which does the translation from `result_internal -> result_external`, but we needed to have a function from `result_external -> BMC_EXIT_CODE`. This is the function that was modelled under the equivalent one from `result_internal -> BMC_EXIT_CODE`.

TGWDB commented on src/goto-bmc/goto_bmc_parse_options.cpp:
```diff
@@ -0,0 +1,100 @@
+  int argc,
+  const char **argv)
+  : parse_options_baset(
+      CORE_CBMC_ONLY_OPTIONS,
```
symex-ready

NlightNFotis replied:
These shouldn't be called `symex-ready` - they refer to the tool not the goto-binaries.

But they will need to be made consistent to the actual tool name, which we settled on `goto-bmc` (this is the name of the tool built). The references are artefacts from a previous change where we called the tool `corebmc` but settled on `goto-bmc` instead.

NlightNFotis replied:
Done.

TGWDB commented on src/goto-bmc/goto_bmc_parse_options.cpp:
```diff
@@ -0,0 +1,100 @@
+      CORE_CBMC_ONLY_OPTIONS,
+      argc,
+      argv,
+      std::string("CORE_CBMC_ONLY") + CBMC_VERSION)
```
symex-ready

NlightNFotis replied:
Renamed to `goto-bmc` - Done.

TGWDB commented on src/goto-bmc/goto_bmc_parse_options.cpp:
```diff
@@ -0,0 +1,100 @@
+  std::cout << api_message_get_string(message) << std::endl;
+}
+
+int core_cbmc_only_parse_optionst::doit()
```
symex-ready

TGWDB commented on src/goto-bmc/goto_bmc_parse_options.cpp:
```diff
@@ -0,0 +1,100 @@
+
+  // With options set, we can now initiate the API and perform analysis.
+  api_sessiont api{api_options};
+  std::cout << "CoreBMC version ";
```
goto-bmc?

NlightNFotis replied:
Done.

TGWDB commented on src/goto-bmc/goto_bmc_parse_options.cpp:
```diff
@@ -0,0 +1,100 @@
+  return verifier_result_to_exit_code(result->final_result());
+}
+
+void core_cbmc_only_parse_optionst::help()
```
symex-ready

TGWDB commented on src/goto-bmc/goto_bmc_parse_options.cpp:
```diff
@@ -0,0 +1,100 @@
+    "goto-bmc [-?] [-h] [--help]      show help\n"
+    "goto-bmc --version               show version and exit\n"
+    // NOLINTNEXTLINE(*)
+    "goto-bmc [options] file.c ...    perform bounded model checking on core goto-binary\n";
```
symex-ready

NlightNFotis replied:
Done.

TGWDB commented on src/goto-bmc/goto_bmc_parse_options.h:
```diff
@@ -0,0 +1,34 @@
+void help();
+
+// clang-format off
+#define CORE_CBMC_ONLY_OPTIONS \
```
symex-ready

TGWDB commented on src/goto-bmc/goto_bmc_parse_options.h:
```diff
@@ -0,0 +1,34 @@
+  "(version)"
+// clang-format on
+
+class core_cbmc_only_parse_optionst : public parse_options_baset
```
symex-ready

TGWDB commented on src/goto-bmc/goto_bmc_parse_options.h:
```diff
@@ -0,0 +1,34 @@
+class core_cbmc_only_parse_optionst : public parse_options_baset
+{
+public:
+  core_cbmc_only_parse_optionst(int argc, const char **argv);
```
symex-ready

TGWDB commented on src/goto-bmc/goto_bmc_parse_options.h:
```diff
@@ -0,0 +1,34 @@
+  virtual int doit();
+};
+
+#endif // CPROVER_CORE_BMC_PARSE_OPTIONS_H
```
symex-ready

NlightNFotis replied:
This is refering to the tool name (`core-bmc`).

TGWDB commented on src/libcprover-cpp/api.h:
```diff
@@ -79,11 +79,28 @@ struct api_sessiont
+  /// @return a `unique_ptr` to the `verification_resultt` summary.
+  std::unique_ptr<verification_resultt> run_verifier() const;
+
+  /// Read a goto-binary from a given filename.
+  /// @warning Will error out if it reads a source file.
+  void read_goto_binary(std::string &file) const;
```
This should check that the goto binary is in symex-ready form. Of course the check is not implemented yet, but maybe this should be a comment or stub in the implementation?

NlightNFotis replied:
We track this functionality, by having a negative test under `regression/goto-bmc/goto-bmc-non-symex-ready-goto` whose job is to fail when fed a non-symex-ready goto.

For now it fails with an invariant violation, so the error message leaves a lot to be desired, but when we fix the recognition of the symex-ready-goto binaries, this test is going to fail (because the invariant violation pattern is not going to be present in the output any more) and this will serve as our reminder to go back and fix that.

TGWDB commented on src/libcprover-cpp/api.h:
```diff
@@ -79,11 +79,28 @@ struct api_sessiont
+  void read_goto_binary(std::string &file) const;
+
+  /// True if file is goto-binary.
+  bool is_goto_binary(std::string &file) const;
```
Maybe here,  -> `is_symex_ready_goto_binary`?

NlightNFotis replied:
We don't want to do this, because we delegate to `is_goto_binary` from `src/goto-programs/read_goto_binary.cpp` which doesn't check the "level" of the goto-binary (whether transformations have been performed and which ones), so we can't really promise to give information we don't have.

TGWDB commented on src/libcprover-cpp/api.h:
```diff
@@ -79,11 +79,28 @@ struct api_sessiont
 private:
   std::unique_ptr<api_session_implementationt> implementation;
+
+  /// Implement necessary transformations to reduce model to core-GOTO, before
```
symex-ready

NlightNFotis replied:
Done.

peterschrammel commented on src/libcprover-cpp/api.cpp:
```diff
@@ -137,6 +137,18 @@ std::unique_ptr<verification_resultt> api_sessiont::verify_model() const
+  return run_verifier();
+}
+
+/// <FILL ME>
```
❓

NlightNFotis replied:
This was empty while we did some refactoring, acting as a reminder to fill in documentation once the code stabilised - it's filled in the next commit.

peterschrammel commented on src/goto-bmc/goto_bmc_parse_options.cpp:
```diff
@@ -0,0 +1,97 @@
+{
+}
+
+void print_messages_to_stdout(
```
❓ What is this function for? 
We should use `log` for printing messages.

NlightNFotis replied:
The way the API works for now is it needs a callback to be passed to it, which it then calls to process messages from it.

This callback function needs to have a specific signature for it to be able to be registered with the API.

What we *can* do however, is change the references to `std::cout` to output things via the `log` instead.

NlightNFotis replied:
Done in 755f0f555721ea372b93f5a1bccef2b6513969c8

thomasspriggs replied:
> Done in [755f0f5](https://github.com/diffblue/cbmc/commit/755f0f555721ea372b93f5a1bccef2b6513969c8)

Is that actually a good idea? With the new API the log exists inside the API layer but is not exposed. So by creating a log in the entry point doesn't that mean two instances of it exist at once?

peterschrammel commented on src/goto-bmc/module_dependencies.txt:
```diff
@@ -0,0 +1,3 @@
+util
+libcprover-cpp
+goto-programs
```
Why this dependency? Shouldn't goto-bmc only depend on libcprover-cpp?

NlightNFotis replied:
We need this because of the check inside `src/goto-bmc/goto_bmc_parse_options.cpp:52` which has a call to `is_goto_binary`, defined under `src/goto-programs/read_goto_binary.cpp`

peterschrammel replied:
If you need it then the API needs to expose it, no?

NlightNFotis replied:
I'm wrong, we already expose it in the API - we will change it to make it defer to the API (we already have this capability, probably lost the idea somewhere during refactoring)

esteffin commented on src/goto-bmc/goto_bmc_parse_options.cpp:
```diff
@@ -0,0 +1,101 @@
+  const api_messaget &message,
+  api_call_back_contextt)
+{
+  // TODO: this is just a hack while a better way to connect the
+  //  parse_options_baset::log to the api callback is found.
+  //  This has to be removed before merging the PR.
+  ui_message_handlert ui_message_handler(cmdlinet{}, "goto-bmc");
+  messaget log{ui_message_handler};
```
This has to be fixed properly.
However the older way to create a log by just calling `messaget log;` was not working.

esteffin replied:
Fixed

esteffin reviewed (changes requested):
The log-handling function has to be fixed properly

thomasspriggs commented on src/libcprover-cpp/api.h:
```diff
@@ -79,11 +79,17 @@ struct api_sessiont
+  /// A simple API version information function.
   std::unique_ptr<std::string> get_api_version() const;
 
+  /// Run the verifier! <FILL ME>
```
:pick: Shouldn't the documentation be included in the commit which adds the function?

thomasspriggs replied:
This looks to be done.

thomasspriggs commented on src/libcprover-cpp/api.cpp:
```diff
@@ -150,11 +162,11 @@ std::unique_ptr<verification_resultt> api_sessiont::verify_model() const
-  {
-    return {};
-  }
+    if(::process_goto_program(
```
:pick: This indenting doesn't look correct. If clang-format complains about this then the corresponding clang-format update should be in this commit and not a later one.

thomasspriggs replied:
This looks to be done.

thomasspriggs commented on src/libcprover-cpp/api.h:
```diff
@@ -79,11 +79,17 @@ struct api_sessiont
   void validate_goto_model() const;
 
-  // A simple API version information function.
+  /// A simple API version information function.
```
:pick: Fix of existing documentation should be in a separate commit from the refactor, if the function is not changed as part of the refactor.

thomasspriggs commented on src/libcprover-cpp/api.h:
```diff
@@ -79,11 +79,17 @@ struct api_sessiont
   std::unique_ptr<api_session_implementationt> implementation;
+
+  /// <FILL ME>
+  bool preprocess_model() const;
```
:bulb: If this function is truly an implementation detail, then it would make sense to me for it to be a free function which is defined the `.cpp` only.
:pick: If the plan is to eventually use this as part of the publicly facing API then it isn't following the existing code pattern, which is to use the call-back to indicate errors and to return a `void` result.

thomasspriggs commented on src/libcprover-cpp/api.cpp:
```diff
@@ -138,15 +139,25 @@ std::unique_ptr<verification_resultt> api_sessiont::verify_model() const
   PRECONDITION(implementation->model);
 
   bool empty_result = preprocess_model();
-  if (empty_result) {
```
:pick: Why is this formatting change in this commit? Comment applies to any other comment on this commit marked with a :foot:

thomasspriggs commented on src/libcprover-cpp/api.cpp:
```diff
@@ -162,11 +173,11 @@ bool api_sessiont::preprocess_model() const
     cprover_c_library_factory);
 
   // Common removal of types and complex constructs
-    if(::process_goto_program(
```
:foot:

thomasspriggs commented on src/libcprover-cpp/api.cpp:
```diff
@@ -183,19 +194,21 @@ bool api_sessiont::preprocess_model() const
   all_properties_verifier_with_trace_storaget<multi_path_symex_checkert>
     verifier(
       *implementation->options, ui_message_handler, *implementation->model);
+
```
:foot:

thomasspriggs commented on src/libcprover-cpp/api.cpp:
```diff
@@ -183,19 +194,21 @@ bool api_sessiont::preprocess_model() const
+
   verification_resultt result;
   auto results = verifier();
+
```
:foot:

thomasspriggs commented on src/libcprover-cpp/api.cpp:
```diff
@@ -183,19 +194,21 @@ bool api_sessiont::preprocess_model() const
   verifier.report();
   // Set results object before returning.
   result.set_result(results);
+
```
:foot:

thomasspriggs commented on src/goto-bmc/goto_bmc_parse_options.cpp:
```diff
@@ -0,0 +1,97 @@
+  api.set_message_callback(print_messages_to_stdout, nullptr);
+
+  // Finally, we read the goto-binary the user supplied...
+  api.read_goto_binary(filename);
```
:no_entry_sign: Missing error handling. I am thinking about handling of basic issues like the specified file not existing. We should not just violate an INVARIANT and dump a stack on bad input from the user. This could be resolved via the message_call back interface via the following steps -

1. Add a `message_levelt` to `api_messaget`.
2. Populate it in `api_message_handlert::print`, using the `level` passed in.
3. Add corresponding getter to the api, which gets the error status from an `api_messaget`.
4. Update the message call back used in this entry point such that the callback context includes an error_encountered flag, which is set to `true` when an error message is sent through the call back.
5. Add exit on error flag tripped.

thomasspriggs replied:
Error handling is still missing here. In the case where `goto_bmc_parse_optionst::doit` calls `api.read_goto_binary(filename);` and the goto binary is not successfully read, control flow should not proceed to the `api.run_verifier();` call below this. If we don't have a good solution for this at the moment, then we should probably put a `TODO` here (and raise a follow-up ticket so it isn't forgotten) at the very least.

thomasspriggs commented on src/goto-bmc/goto_bmc_parse_options.h:
```diff
@@ -0,0 +1,34 @@
+#include <string>
+#include <vector>
+
+api_optionst parse_command_line_options(const std::vector<std::string> &);
```
:no_entry_sign: This `parse_command_line_options` function seems to be unimplemented at this point in the commit history. The same comment applies to `void help();` Therefore the forward declaration should not exist in this commit.

thomasspriggs replied:
This looks to be done.

thomasspriggs commented on src/goto-bmc/goto_bmc_parse_options.h:
```diff
@@ -0,0 +1,34 @@
+public:
+  goto_bmc_parse_optionst(int argc, const char **argv);
+
+  virtual void help();
```
:hammer: As we are overriding the `help` function from `parse_options_baset`, this should be specified as `void help() override;` without `virtual`. Note that this is standardised in `CODING_STANDARD.md` line 196.
This comment also applies to the `doit` function just below.

thomasspriggs replied:
This looks to be done.

thomasspriggs commented on src/goto-bmc/goto_bmc_parse_options.cpp:
```diff
@@ -0,0 +1,97 @@
+    api_options.validate_goto_model(true);
+  }
+
+  // With options set, we can now initiate the API and perform analysis.
```
:no_entry_sign: This comment is misleading, because the block of code immediately below it does not perform the analysis. 
:pick: Also is a comment which says "With options set, we can now initiate the API" useful, when the line below it is `api_sessiont api{api_options};` which says initialise the API session using the options (in C++)? IMO the code should tell you what the code does. With well named code there should be no need to translate C++ to English and add the English to a comment. It just leaves scope for the English to mismatch the C++. Comments are really valuable when they explain _**why**_ the code is written the way it is, not just existing as a duplication of what the reader can see for themselves.

NlightNFotis replied:
Yeah, that has been removed, but I believe in a commit later than the one it was commented on. Agree with the comment in the review window (the original commits are there to guide our implementation during the green-field implementation, and they fell through the cracks during various refactorings/handoffs of the code).

thomasspriggs replied:
This looks to be resolved.

thomasspriggs commented on src/goto-bmc/goto_bmc_parse_options.cpp:
```diff
@@ -0,0 +1,97 @@
+  std::cout << *api.get_api_version() << std::endl;
+
+  // The API works with a callback for querying state - we for now just print
+  // collected messages from the message buffer to stdout.
```
:information_source: A capture-less lambda can be converted to a function pointer (without using `std::function`). This technique could be used in order to keep the definition of the function next-to/in-line-with the setting of the message callback. This would restrict the scope of the function and might dodge such reader questions as "what is this function for?"

thomasspriggs commented on src/libcprover-cpp/verification_result.h:
```diff
@@ -80,4 +80,8 @@ struct verification_resultt
   std::unique_ptr<verification_result_implt> _impl;
 };
 
+// Allow translation of `verifier_resultt` into a CPROVER_EXIT_CODES (so that
```
:pick: The use of a plural `CPROVER_EXIT_CODES` suggests one result is turned into multiple codes.
:thinking: Given that the API headers may be distributed without the rest of the project source, the reader maybe unaware of the existence of `src/util/exit_codes.h`. So maybe we should just call this a Command Line Interface (CLI) return value/exit code; especially given that the result is not strongly typed.

thomasspriggs commented on src/libcprover-cpp/api.cpp:
```diff
@@ -137,6 +138,28 @@ std::unique_ptr<verification_resultt> api_sessiont::verify_model() const
+void api_sessiont::read_goto_binary(std::string &file) const
+{
+  implementation->model = util_make_unique<goto_modelt>(
+    ::read_goto_binary(file, *implementation->message_handler).value());
```
:no_entry_sign: `read_goto_binary` may return an empty optional. In that case this implementation will call `.value()` on it to get the non-existent value; which isn't exactly going to fail gracefully.

NlightNFotis replied:
Done (error handling added in the first commit)

thomasspriggs commented on src/libcprover-cpp/api.cpp:
```diff
@@ -137,6 +138,28 @@ std::unique_ptr<verification_resultt> api_sessiont::verify_model() const
+    ::read_goto_binary(file, *implementation->message_handler).value());
+}
+
+bool api_sessiont::is_goto_binary(std::string &file) const
```
:grey_question: Please consider whether having a separate `read_goto_binary` function and a `is_goto_binary` function is actually a good interface for a publicly facing API? This kind of implies that the file has to be opened and closed twice for the typical use case.

NlightNFotis replied:
This requires some more thinking as to how it's positioned in the API, but the `is_goto_binary` opens the file and reads just 8 bytes of it (a header) to determine whether it's a goto-binary, so it doesn't result in much inefficiency (if I understand your concerns).

From a code organisation standpoint, yeah, I need to think about it a bit more.

thomasspriggs replied:
Releasing the file resource between the two API calls has potential concurrency implications. Other system processes have the opportunity to perform various operations in between the call to `is_goto_binary` and `read_goto_binary` which could make the second call fail. I admit this isn't hugely likely, but the possibility of these kinds of race conditions seems worth consideration to me.

NlightNFotis replied:
We know that CBMC is not thread-safe in general, so I admit that this is a possibility, but bar from starting to add locks to various calls like that right now I don't know how to fix it.

I believe that adding locks/making it thread-safe probably falls outside of the scope of this PR. Do you have any suggestions on another way to achieve this, or did you want locks introduced?

thomasspriggs replied:
Thread safety is entirely unrelated issue here. Processes contain threads, the cbmc process has process-level (global) state which would be an issue if we wanted to use multiple threads within the cbmc process. This isn't what this issue is about, as cbmc does not use multiple threads and there is no plan to introduce them to my knowledge. However we should correctly support multi-processing (play-nicely) whether the processes involved are multiple instances of cbmc, or cbmc plus other unrelated user processes. I could go into hypothetical scenarios about how non-related processes could cause issues with the goto-binary in between the API calls but this isn't really the crux of the issue.

I think that having two separate API calls invites incorrect assumptions about how the two API calls relate to each other. Simply removing `is_goto_binary` from the public API would be one way to address this. Potential incorrect assumptions include -

- The assumption that if `is_goto_binary` returns true then the following call to `read_goto_binary` will succeed. This is incorrect even without considering concurrency / race condition scenarios. `is_goto_binary` only checks the header of the file. Therefore `read_goto_binary` could still fail due to error causing data after the header. Therefore the caller of `read_goto_binary` must handle any errors encountered during the API call even if a preceding call to `is_goto_binary` suceeded.
- The assumption that if we are going to call `read_goto_binary` then we must call `is_goto_binary` before hand. As I read the current implementation, `read_goto_binary` includes all the checks which `is_goto_binary` does. Therefore doing a precheck on `is_goto_binary` is unnecessary as it is redundant.

We could keep a publicly facing API call which does a check on whether a particular file is a goto-binary or not. But I don't see what the current advantage of doing so would be?

NlightNFotis replied:
Ah, I think I understand the problem better now. I will paraphrase the concern, to make sure I have understood it correctly - feel free to correct any points that are wrong:

> The issue with `is_goto_binary` and `read_goto_binary` is that the similar naming
> confuses because it implies an implicit sequencing between the two calls that sets
> up certain expectations that can be violated.

If my understanding is correct, would some documentation ameliorate those concerns?

With regards to the API call itself, I agree that it looks a bit out of place, but in general, my opinion is that the API as a whole should provide mechanism, not policy. I can easily picture the API call being part of some future scenarios, something like (pseudocode)

```c
if (is_goto_binary())
   extract_goto_binary_section(relro)
else if (is_elf_binary())
   extract_elf_binary_section(retro)
else
   // don't do anything for source files
```

I don't know how realistic this particular example might be, but scenarios that require some sort of querying of `goto-binary` properties fall within the purview of general API usage (if a bit underdeveloped right now), so I wouldn't want to remove from that capability.

Any thoughts?

thomasspriggs replied:
I think that we should shelve this discussion for potential continuation outside the forum of this PR review, as it was a point for consideration rather than a blocking comment.

thomasspriggs commented on src/goto-bmc/goto_bmc_parse_options.cpp:
```diff
@@ -48,7 +48,7 @@ int goto_bmc_parse_optionst::doit()
 
   if(
-    cmdline.args.size() != 1 ||
+    cmdline.args.size() > 1 &&
```
This change appears to be in the wrong commit, given that the commit title refers to regression tests, but this change is a change to the implementation.

thomasspriggs replied:
This appears to be resolved.

thomasspriggs commented on regression/goto-bmc/goto-bmc-non-symex-ready-goto/negative1/main.desc:
```diff
@@ -1,7 +1,7 @@
 CORE
 main.c
 
-^EXIT=[127|134]$
```
:pick: The commit title refers to moving tests, but this part is changing the test.

esteffin reviewed (approved):
Change requested has been fulfilled

thomasspriggs committed:
Refactor call back function to accept a struct context

In preparation for expanding the context beyond just the message stream.

thomasspriggs committed:
Add file read error handling to `goto_bmc`

This commit adds exiting on bad input files as is proper, instead of
proceeding to further worse failures such as INVARIANT violations or
even seg faults.

thomasspriggs committed:
Expand options pack in `symex-ready-goto/chain.sh`

This would previously been left as a single unexpanded argument
regardless of the number of arguments in the `.desc` file. For example
`goto_bmc main.goto.symex_ready ''` in the case of no arguments
specified. This should be expanded to match the number of arguments in
the `options` variable. For example `goto_bmc main.goto.symex_ready` in
the case of no arguments or `goto_bmc main.goto.symex_ready --argument1
--argument2` in the case of two arguments in the desc file.

thomasspriggs committed:
Fix argument check

Zero arguments should trigger an error as should files which fail the
is_goto_binary check when a single file is specified.

thomasspriggs committed:
Add tests of error handling on reading invalid goto binaries

thomasspriggs committed:
Add tests for incorrect number of goto binaries specified

thomasspriggs committed:
Add missing end of message in `goto_bmc_parse_optionst::help`

Running `goto-bmc --help` was previously resulting in the tool exiting
without displaying help text. This was due to the incomplete message
being cached, but never sent to the console output before exiting.

thomasspriggs committed:
Add `--version` command line argument test

thomasspriggs committed:
Add `--help` command line argument test

thomasspriggs committed:
Remove duplicated `--help` argument handling

The `parse_options_baset::main` function checks for a `--help` argument.
It calls the `help()` member function and exits rather than calling
`doit()` when this argument is specified. The duplicated code removed by
this commit was misleading because it was never reached when the
`--help` argument was specified.

thomasspriggs reviewed (approved):
The missing error handling has now been added.

Commit:
Set code owners for `src/goto-bmc`.

Setting the `CODEOWNERS` file entry for `src/goto-bmc` folder.

thomasspriggs commented on CODEOWNERS:
```diff
@@ -50,6 +50,7 @@
 # These files change frequently and changes are medium-risk
 
 /src/goto-analyzer/ @martin-cs @peterschrammel
+/src/goto-bmc/ @NlightNFotis @thomasspriggs @esteffin @TGWDB @peterschrammel
```
:+1:
